client-go/src/main/client/session.go

```diff
@@ -0,0 +1,667 @@
+	}
+	iProtocol := protocolFactory.GetProtocol(s.trans)
+	oProtocol := protocolFactory.GetProtocol(s.trans)
+	s.client = rpc.NewTSIServiceClient(thrift.NewTStandardClient(iProtocol, oProtocol))
+	s.ZoneId = DefaultZoneId
+	tSOpenSessionReq := rpc.TSOpenSessionReq{ClientProtocol: protocolVersion, ZoneId: s.ZoneId, Username: &s.User,
+		Password: &s.Passwd}
+	tSOpenSessionResp, err := s.client.OpenSession(context.Background(), &tSOpenSessionReq)
+	if err != nil {
+		log.WithError(err).Error("open session failed")
+	} else {
+		log.WithField("code", tSOpenSessionResp.GetStatus().Code).Debug("open session success")
+	}
+	s.sessionId = tSOpenSessionResp.GetSessionId()
+	s.requestStatementId, err = s.client.RequestStatementId(context.Background(), s.sessionId)
+	s.FetchSize = DefaultFetchSize
+	if err != nil {
+		log.WithError(err).Error("request StatementId failed")
+	}
+	if s.ZoneId != "" {
+		s.SetTimeZone(s.ZoneId)
+	} else {
+		s.ZoneId = s.GetTimeZone()
+	}
+}
+
+func (s *Session) CheckTimeseriesExists(path string) bool {
+	dataSet := s.ExecuteQueryStatement("SHOW TIMESERIES " + path)
+	result := dataSet.HasNext()
+	dataSet.CloseOperationHandle()
+	return result
+}
+
+func (s *Session) Close() {
+	tSCloseSessionReq := rpc.NewTSCloseSessionReq()
+	tSCloseSessionReq.SessionId = s.sessionId
+	status, err := s.client.CloseSession(context.Background(), tSCloseSessionReq)
+	if err != nil {
+		log.WithError(err).Error("Error occurs when closing session at server. Maybe server is down.")
+	} else {
+		log.WithField("code", status.Code).Debug("close session success")
+	}
+	s.trans.Close()
+}
+
+/*
+ *set one storage group
+ *
+ *param
+ *storageGroupId: string, storage group name (starts from root)
+ *
+ */
+func (s *Session) SetStorageGroup(storageGroupId string) {
+	status, err := s.client.SetStorageGroup(context.Background(), s.sessionId, storageGroupId)
+	if err != nil {
+		log.WithError(err).Error("setting storage group failed")
+	} else {
+		log.WithFields(log.Fields{
+			"sg":   storageGroupId,
+			"code": status.Code,
+		}).Debug("setting storage group success")
+	}
+}
+
+/*
+ *delete one storage group
+ *
+ *param
+ *storageGroupId: string, storage group name (starts from root)
+ *
+ */
+func (s *Session) DeleteStorageGroup(storageGroupId string) {
+	status, err := s.client.DeleteStorageGroups(context.Background(), s.sessionId, []string{storageGroupId})
+	if err != nil {
+		log.WithError(err).Error("delete storage group failed")
+	} else {
+		log.WithFields(log.Fields{
+			"sg":   storageGroupId,
+			"code": status.Code,
+		}).Debug("delete storage group success")
+	}
+}
+
+/*
+ *delete multiple storage group
+ *
+ *param
+ *storageGroupIds: []string, paths of the target storage groups
+ *
+ */
+func (s *Session) DeleteStorageGroups(storageGroupIds []string) {
+	status, err := s.client.DeleteStorageGroups(context.Background(), s.sessionId, storageGroupIds)
+	s.sg = strings.Replace(strings.Trim(fmt.Sprint(storageGroupIds), "[]"), " ", ",", -1)
```

Comment:
Do not do the code here as you just want to print the devices in debug log, for every operation, the code will be called, it'll cost cpu
`s.sg = strings.Replace(strings.Trim(fmt.Sprint(storageGroupIds), "[]"), " ", ",", -1)`

Reply:
thanks for your guidance